### ISSUE TITLE
Lots of errors when building xgboost on Windows 10

### ISSUE THREAD

**Issue by llmercury** [dmlc/xgboost#2260]
## Environment info
**Operating System**: Windows 10 Home 64 bit

**Compiler**: Mingw-w64, x86_64-7.1.0-posix-seh-rt_v5-rev0

**Package used**: python 3.6, Anaconda 4.3.1 64-bit (just downloaded from https://www.continuum.io/downloads )

**`xgboost` version used**: Not sure, v0.6? It should be the newest version. I used "git clone --recursive https://github.com/dmlc/xgboost" to get it.

I followed the installation guide at http://xgboost.readthedocs.io/en/latest/build.html

In a git bash, I did:
git clone --recursive https://github.com/dmlc/xgboost
git submodule init
git submodule update
alias make='mingw32-make'
cd /c/Users/m/code/xgboost
cp make/mingw64.mk config.mk; make -j4

I got the message:

g++ -m64 -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude -DDMLC_ENABLE_STD_THREAD=0  -Idmlc-core/include -Irabit/include -O3 -funroll-loops -msse2 -fopenmp -MM -MT build/data/sparse_page_raw_format.o src/data/sparse_page_raw_format.cc >build/data/sparse_page_raw_format.d
g++ -m64 -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude -DDMLC_ENABLE_STD_THREAD=0  -Idmlc-core/include -Irabit/include -O3 -funroll-loops -msse2 -fopenmp -MM -MT build/data/simple_dmatrix.o src/data/simple_dmatrix.cc >build/data/simple_dmatrix.d
g++ -m64 -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude -DDMLC_ENABLE_STD_THREAD=0  -Idmlc-core/include -Irabit/include -O3 -funroll-loops -msse2 -fopenmp -MM -MT build/data/data.o src/data/data.cc >build/data/data.d
g++ -m64 -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude -DDMLC_ENABLE_STD_THREAD=0  -Idmlc-core/include -Irabit/include -O3 -funroll-loops -msse2 -fopenmp -MM -MT build/data/simple_csr_source.o src/data/simple_csr_source.cc >build/data/simple_csr_source.d
g++ -m64 -c -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude -DDMLC_ENABLE_STD_THREAD=0  -Idmlc-core/include -Irabit/include -O3 -funroll-loops -msse2 -fopenmp src/data/simple_csr_source.cc -o build/data/simple_csr_source.o
g++ -m64 -c -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude -DDMLC_ENABLE_STD_THREAD=0  -Idmlc-core/include -Irabit/include -O3 -funroll-loops -msse2 -fopenmp src/data/sparse_page_raw_format.cc -o build/data/sparse_page_raw_format.o
g++ -m64 -c -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude -DDMLC_ENABLE_STD_THREAD=0  -Idmlc-core/include -Irabit/include -O3 -funroll-loops -msse2 -fopenmp src/data/data.cc -o build/data/data.o
g++ -m64 -c -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude -DDMLC_ENABLE_STD_THREAD=0  -Idmlc-core/include -Irabit/include -O3 -funroll-loops -msse2 -fopenmp src/data/simple_dmatrix.cc -o build/data/simple_dmatrix.o
In file included from src/data/data.cc:9:0:
src/data/./sparse_batch_page.h:238:46: error: 'function' is not a member of 'std'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
In file included from src/data/sparse_page_raw_format.cc:8:0:
src/data/./sparse_batch_page.h:238:46: error: 'function' is not a member of 'std'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
src/data/./sparse_batch_page.h:238:46: note: suggested alternative: 'is_function'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
                                              is_function
src/data/./sparse_batch_page.h:238:46: error: 'function' is not a member of 'std'
src/data/./sparse_batch_page.h:238:46: note: suggested alternative: 'is_function'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
                                              is_function
src/data/./sparse_batch_page.h:238:77: error: template argument 2 is invalid
                                         std::function<SparsePage::Format* ()> > {
                                                                             ^
src/data/./sparse_batch_page.h:238:79: error: expected '{' before '>' token
                                         std::function<SparsePage::Format* ()> > {
                                                                               ^
src/data/./sparse_batch_page.h:238:79: error: expected unqualified-id before '>' token
src/data/./sparse_batch_page.h:238:46: note: suggested alternative: 'is_function'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
                                              is_function
src/data/./sparse_batch_page.h:238:46: error: 'function' is not a member of 'std'
src/data/./sparse_batch_page.h:238:46: note: suggested alternative: 'is_function'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
                                              is_function
src/data/./sparse_batch_page.h:238:77: error: template argument 2 is invalid
                                         std::function<SparsePage::Format* ()> > {
                                                                             ^
src/data/./sparse_batch_page.h:238:79: error: expected '{' before '>' token
                                         std::function<SparsePage::Format* ()> > {
                                                                               ^
src/data/./sparse_batch_page.h:238:79: error: expected unqualified-id before '>' token
In file included from dmlc-core/include/dmlc/././parameter.h:25:0,
                 from dmlc-core/include/dmlc/./registry.h:14,
                 from dmlc-core/include/dmlc/data.h:16,
                 from include/xgboost/data.h:11,
                 from src/data/./simple_csr_source.h:12,
                 from src/data/simple_csr_source.cc:7:
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'dmlc::optional<T>::~optional() [with T = int]':
dmlc-core/include/dmlc/././parameter.h:513:7:   required from here
dmlc-core/include/dmlc/./././optional.h:60:7: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
       reinterpret_cast<T*>(&val)->~T();
       ^~~~~~~~~~~~~~~~
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'const T& dmlc::optional<T>::value() const [with T = int]':
dmlc-core/include/dmlc/././parameter.h:872:9:   required from here
dmlc-core/include/dmlc/./././optional.h:106:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return *reinterpret_cast<const T*>(&val);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/data.h:16:0,
                 from include/xgboost/data.h:11,
                 from src/data/sparse_page_raw_format.cc:6:
dmlc-core/include/dmlc/./registry.h:247:55: error: invalid use of incomplete type 'struct xgboost::data::SparsePageFormatReg'
       ::dmlc::Registry<EntryType>::Get()->__REGISTER__(#Name)           \
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
src/data/./sparse_batch_page.h:254:3: note: in expansion of macro 'DMLC_REGISTRY_REGISTER'
   DMLC_REGISTRY_REGISTER(::xgboost::data::SparsePageFormatReg, SparsePageFormat, Name)
   ^~~~~~~~~~~~~~~~~~~~~~
src/data/sparse_page_raw_format.cc:93:1: note: in expansion of macro 'XGBOOST_REGISTER_SPARSE_PAGE_FORMAT'
 XGBOOST_REGISTER_SPARSE_PAGE_FORMAT(raw)
 ^
In file included from src/data/sparse_page_raw_format.cc:8:0:
src/data/./sparse_batch_page.h:236:8: note: forward declaration of 'struct xgboost::data::SparsePageFormatReg'
 struct SparsePageFormatReg
        ^~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/data.h:16:0,
                 from include/xgboost/data.h:11,
                 from src/data/sparse_page_raw_format.cc:6:
dmlc-core/include/dmlc/./registry.h:247:55: error: invalid use of incomplete type 'struct xgboost::data::SparsePageFormatReg'
       ::dmlc::Registry<EntryType>::Get()->__REGISTER__(#Name)           \
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
src/data/./sparse_batch_page.h:254:3: note: in expansion of macro 'DMLC_REGISTRY_REGISTER'
   DMLC_REGISTRY_REGISTER(::xgboost::data::SparsePageFormatReg, SparsePageFormat, Name)
   ^~~~~~~~~~~~~~~~~~~~~~
src/data/sparse_page_raw_format.cc:93:1: note: in expansion of macro 'XGBOOST_REGISTER_SPARSE_PAGE_FORMAT'
 XGBOOST_REGISTER_SPARSE_PAGE_FORMAT(raw)
 ^
In file included from src/data/sparse_page_raw_format.cc:8:0:
src/data/./sparse_batch_page.h:236:8: note: forward declaration of 'struct xgboost::data::SparsePageFormatReg'
 struct SparsePageFormatReg
        ^~~~~~~~~~~~~~~~~~~
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'const T& dmlc::optional<T>::operator*() const [with T = int]':
dmlc-core/include/dmlc/./././optional.h:134:11:   required from 'std::ostream& dmlc::operator<<(std::ostream&, const dmlc::optional<T>&) [with T = int; std::ostream = std::basic_ostream<char>]'
dmlc-core/include/dmlc/././parameter.h:877:13:   required from here
dmlc-core/include/dmlc/./././optional.h:97:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return *reinterpret_cast<const T*>(&val);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/././parameter.h:25:0,
                 from dmlc-core/include/dmlc/./registry.h:14,
                 from dmlc-core/include/dmlc/data.h:16,
                 from include/xgboost/data.h:11,
                 from src/data/sparse_page_raw_format.cc:6:
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'dmlc::optional<T>::~optional() [with T = int]':
dmlc-core/include/dmlc/././parameter.h:513:7:   required from here
dmlc-core/include/dmlc/./././optional.h:60:7: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
       reinterpret_cast<T*>(&val)->~T();
       ^~~~~~~~~~~~~~~~
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'const T& dmlc::optional<T>::value() const [with T = int]':
dmlc-core/include/dmlc/././parameter.h:872:9:   required from here
dmlc-core/include/dmlc/./././optional.h:106:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return *reinterpret_cast<const T*>(&val);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/data.h:16:0,
                 from include/xgboost/data.h:11,
                 from src/data/sparse_page_raw_format.cc:6:
dmlc-core/include/dmlc/./registry.h: In instantiation of 'EntryType& dmlc::Registry<EntryType>::__REGISTER__(const string&) [with EntryType = xgboost::data::SparsePageFormatReg; std::__cxx11::string = std::__cxx11::basic_string<char>]':
src/data/sparse_page_raw_format.cc:93:1:   required from here
dmlc-core/include/dmlc/./registry.h:80:20: error: invalid use of incomplete type 'struct xgboost::data::SparsePageFormatReg'
     EntryType *e = new EntryType();
                    ^~~~~~~~~~~~~~~
In file included from src/data/sparse_page_raw_format.cc:8:0:
src/data/./sparse_batch_page.h:236:8: note: forward declaration of 'struct xgboost::data::SparsePageFormatReg'
 struct SparsePageFormatReg
        ^~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/data.h:16:0,
                 from include/xgboost/data.h:11,
                 from src/data/sparse_page_raw_format.cc:6:
dmlc-core/include/dmlc/./registry.h:81:8: error: invalid use of incomplete type 'struct xgboost::data::SparsePageFormatReg'
     e->name = name;
     ~~~^~~~
In file included from src/data/sparse_page_raw_format.cc:8:0:
src/data/./sparse_batch_page.h:236:8: note: forward declaration of 'struct xgboost::data::SparsePageFormatReg'
 struct SparsePageFormatReg
        ^~~~~~~~~~~~~~~~~~~
src/data/data.cc: In static member function 'static xgboost::data::SparsePage::Format* xgboost::data::SparsePage::Format::Create(const string&)':
src/data/data.cc:281:12: error: invalid use of incomplete type 'const struct xgboost::data::SparsePageFormatReg'
   return (e->body)();
            ^~
In file included from src/data/data.cc:9:0:
src/data/./sparse_batch_page.h:236:8: note: forward declaration of 'struct xgboost::data::SparsePageFormatReg'
 struct SparsePageFormatReg
        ^~~~~~~~~~~~~~~~~~~
src/data/data.cc:281:14: error: invalid use of incomplete type 'const struct xgboost::data::SparsePageFormatReg'
   return (e->body)();
              ^~~~
In file included from src/data/data.cc:9:0:
src/data/./sparse_batch_page.h:236:8: note: forward declaration of 'struct xgboost::data::SparsePageFormatReg'
 struct SparsePageFormatReg
        ^~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/././parameter.h:25:0,
                 from dmlc-core/include/dmlc/./registry.h:14,
                 from dmlc-core/include/dmlc/data.h:16,
                 from include/xgboost/data.h:11,
                 from src/data/sparse_page_raw_format.cc:6:
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'const T& dmlc::optional<T>::operator*() const [with T = int]':
dmlc-core/include/dmlc/./././optional.h:134:11:   required from 'std::ostream& dmlc::operator<<(std::ostream&, const dmlc::optional<T>&) [with T = int; std::ostream = std::basic_ostream<char>]'
dmlc-core/include/dmlc/././parameter.h:877:13:   required from here
dmlc-core/include/dmlc/./././optional.h:97:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return *reinterpret_cast<const T*>(&val);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/././parameter.h:25:0,
                 from dmlc-core/include/dmlc/./registry.h:14,
                 from dmlc-core/include/dmlc/data.h:16,
                 from include/xgboost/data.h:11,
                 from src/data/data.cc:5:
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'dmlc::optional<T>::~optional() [with T = int]':
dmlc-core/include/dmlc/././parameter.h:513:7:   required from here
dmlc-core/include/dmlc/./././optional.h:60:7: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
       reinterpret_cast<T*>(&val)->~T();
       ^~~~~~~~~~~~~~~~
In file included from src/data/./simple_dmatrix.h:15:0,
                 from src/data/simple_dmatrix.cc:11:
src/data/././sparse_batch_page.h:238:46: error: 'function' is not a member of 'std'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
src/data/././sparse_batch_page.h:238:46: note: suggested alternative: 'is_function'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
                                              is_function
src/data/././sparse_batch_page.h:238:46: error: 'function' is not a member of 'std'
src/data/././sparse_batch_page.h:238:46: note: suggested alternative: 'is_function'
                                         std::function<SparsePage::Format* ()> > {
                                              ^~~~~~~~
                                              is_function
src/data/././sparse_batch_page.h:238:77: error: template argument 2 is invalid
                                         std::function<SparsePage::Format* ()> > {
                                                                             ^
src/data/././sparse_batch_page.h:238:79: error: expected '{' before '>' token
                                         std::function<SparsePage::Format* ()> > {
                                                                               ^
src/data/././sparse_batch_page.h:238:79: error: expected unqualified-id before '>' token
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'const T& dmlc::optional<T>::value() const [with T = int]':
dmlc-core/include/dmlc/././parameter.h:872:9:   required from here
dmlc-core/include/dmlc/./././optional.h:106:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return *reinterpret_cast<const T*>(&val);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/data.h:16:0,
                 from include/xgboost/data.h:11,
                 from src/data/data.cc:5:
dmlc-core/include/dmlc/./registry.h: In instantiation of 'dmlc::Registry<EntryType>::~Registry() [with EntryType = xgboost::data::SparsePageFormatReg]':
src/data/data.cc:21:1:   required from here
dmlc-core/include/dmlc/./registry.h:118:7: warning: possible problem detected in invocation of delete operator: [-Wdelete-incomplete]
       delete entry_list_[i];
       ^~~~~~
dmlc-core/include/dmlc/./registry.h:118:7: warning: invalid use of incomplete type 'struct xgboost::data::SparsePageFormatReg'
In file included from src/data/data.cc:9:0:
src/data/./sparse_batch_page.h:236:8: note: forward declaration of 'struct xgboost::data::SparsePageFormatReg'
 struct SparsePageFormatReg
        ^~~~~~~~~~~~~~~~~~~
In file included from dmlc-core/include/dmlc/data.h:16:0,
                 from include/xgboost/data.h:11,
                 from src/data/data.cc:5:
dmlc-core/include/dmlc/./registry.h:118:7: note: neither the destructor nor the class-specific operator delete will be called, even if they are declared when the class is defined
       delete entry_list_[i];
       ^~~~~~
In file included from dmlc-core/include/dmlc/././parameter.h:25:0,
                 from dmlc-core/include/dmlc/./registry.h:14,
                 from dmlc-core/include/dmlc/data.h:16,
                 from include/xgboost/data.h:11,
                 from src/data/data.cc:5:
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'const T& dmlc::optional<T>::operator*() const [with T = int]':
dmlc-core/include/dmlc/./././optional.h:134:11:   required from 'std::ostream& dmlc::operator<<(std::ostream&, const dmlc::optional<T>&) [with T = int; std::ostream = std::basic_ostream<char>]'
dmlc-core/include/dmlc/././parameter.h:877:13:   required from here
dmlc-core/include/dmlc/./././optional.h:97:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return *reinterpret_cast<const T*>(&val);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Makefile:117: recipe for target 'build/data/sparse_page_raw_format.o' failed
mingw32-make: *** [build/data/sparse_page_raw_format.o] Error 1
mingw32-make: *** Waiting for unfinished jobs....
In file included from dmlc-core/include/dmlc/././parameter.h:25:0,
                 from dmlc-core/include/dmlc/./registry.h:14,
                 from dmlc-core/include/dmlc/data.h:16,
                 from include/xgboost/data.h:11,
                 from src/data/simple_dmatrix.cc:7:
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'dmlc::optional<T>::~optional() [with T = int]':
dmlc-core/include/dmlc/././parameter.h:513:7:   required from here
dmlc-core/include/dmlc/./././optional.h:60:7: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
       reinterpret_cast<T*>(&val)->~T();
       ^~~~~~~~~~~~~~~~
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'const T& dmlc::optional<T>::value() const [with T = int]':
dmlc-core/include/dmlc/././parameter.h:872:9:   required from here
dmlc-core/include/dmlc/./././optional.h:106:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return *reinterpret_cast<const T*>(&val);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dmlc-core/include/dmlc/./././optional.h: In instantiation of 'const T& dmlc::optional<T>::operator*() const [with T = int]':
dmlc-core/include/dmlc/./././optional.h:134:11:   required from 'std::ostream& dmlc::operator<<(std::ostream&, const dmlc::optional<T>&) [with T = int; std::ostream = std::basic_ostream<char>]'
dmlc-core/include/dmlc/././parameter.h:877:13:   required from here
dmlc-core/include/dmlc/./././optional.h:97:12: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return *reinterpret_cast<const T*>(&val);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Makefile:117: recipe for target 'build/data/data.o' failed
mingw32-make: *** [build/data/data.o] Error 1
Makefile:117: recipe for target 'build/data/simple_dmatrix.o' failed
mingw32-make: *** [build/data/simple_dmatrix.o] Error 1

I also tried the steps at https://www.ibm.com/developerworks/community/blogs/jfp/entry/Installing_XGBoost_For_Anaconda_on_Windows?lang=ru

added:
$ cd dmlc-core
$ make -j4
$ cd ../rabit
$ make lib/librabit_empty.a -j4
$ cd ..

before:
$ cp make/mingw64.mk config.mk
$ make -j4

Still, the same errors.

Please help!

**Comment by AbduGadir** [comment]
having exactly the same problem

**Comment by superbobry** [comment]
Looking at the errors, I think they should be fixed by #2256. Could you try it our and lmk about the results?

**Comment by nshawen** [comment]
Was having the same issues on Windows 7. I tried out #2256 (along with the fix to '/src/data/sparse_batch_page.h' suggested by ClimberPG) and was able to build successfully.
